File: python_backend/app/models/execution.py
"""
統一的執行記錄資料模型

此模組定義了 TWAP 和 Pairs 策略的統一執行記錄格式，
確保前端和數據分析只需處理一種格式。
"""

from typing import Optional, Literal
from pydantic import BaseModel, Field
import time


class ExecutionLeg(BaseModel):
    """執行腿的統一結構"""
    exchange: str = Field(description="交易所名稱")
    symbol: str = Field(description="交易對符號")
    type: Literal["spot", "linear", "inverse"] = Field(description="交易類型")
    side: Literal["buy", "sell"] = Field(description="買賣方向")
    orderId: Optional[str] = Field(default=None, description="訂單ID")
    price: Optional[float] = Field(default=None, description="成交價格")
    priceUpdated: bool = Field(default=False, description="價格是否已更新")
    originalOrderId: Optional[str] = Field(default=None, description="原始訂單ID（回滾時使用）")

    class Config:
        json_schema_extra = {
            "example": {
                "exchange": "bybit",
                "symbol": "ETHUSDT",
                "type": "linear",
                "side": "buy",
                "orderId": "abc123",
                "price": 3000.0,
                "priceUpdated": True,
                "originalOrderId": None
            }
        }


class UnifiedExecutionRecord(BaseModel):
    """統一的執行記錄格式（TWAP 和 Pairs 共用）"""
    
    # 基本資訊
    ts: int = Field(description="時間戳（毫秒）")
    mode: Literal["pair", "twap"] = Field(description="策略類型")
    strategyId: str = Field(description="策略ID（統一識別用）")
    pairId: Optional[str] = Field(default=None, description="Pairs策略ID（mode=pair時使用）")
    twapId: Optional[str] = Field(default=None, description="TWAP策略ID（mode=twap時使用）")
    
    # 執行資訊
    totalTriggers: int = Field(ge=1, description="第幾次執行（從1開始，pair和twap共用）")
    
    # 狀態資訊
    status: Literal["success", "failed", "cancelled", "rolled_back"] = Field(description="執行狀態")
    reason: Optional[str] = Field(default=None, description="失敗或取消原因")
    error: Optional[str] = Field(default=None, description="錯誤訊息")
    
    # 數量和價格資訊
    qty: float = Field(gt=0, description="本次執行數量")
    spread: Optional[float] = Field(default=None, description="價差")
    spreadPercent: Optional[float] = Field(default=None, description="價差百分比")
    
    # 策略配置資訊
    totalAmount: float = Field(gt=0, description="預期總數量")
    orderCount: int = Field(ge=1, description="總執行次數/最大執行次數")
    threshold: Optional[float] = Field(default=None, description="觸發閾值（pair用，twap為null）")
    intervalMs: Optional[int] = Field(default=None, description="執行間隔毫秒（twap用，pair為null）")
    
    # 回滾資訊
    isRollback: bool = Field(default=False, description="是否為回滾記錄")
    
    # 執行腿資訊
    leg1: Optional[ExecutionLeg] = Field(default=None, description="第一腿執行結果")
    leg2: Optional[ExecutionLeg] = Field(default=None, description="第二腿執行結果")

    class Config:
        json_schema_extra = {
            "examples": [
                {
                    "description": "Pairs 執行記錄範例",
                    "value": {
                        "ts": 1766508230641,
                        "mode": "pair",
                        "strategyId": "pair_1766508229326_nd8p9x",
                        "pairId": "pair_1766508229326_nd8p9x",
                        "twapId": None,
                        "totalTriggers": 1,
                        "status": "success",
                        "reason": None,
                        "error": None,
                        "qty": 0.01,
                        "spread": -0.03,
                        "spreadPercent": -0.001016,
                        "totalAmount": 0.01,
                        "orderCount": 1,
                        "threshold": -1.0,
                        "intervalMs": None,
                        "isRollback": False,
                        "leg1": {
                            "exchange": "bybit",
                            "symbol": "ETHUSDT",
                            "type": "linear",
                            "side": "sell",
                            "orderId": "abc123",
                            "price": 2952.96,
                            "priceUpdated": True,
                            "originalOrderId": None
                        },
                        "leg2": {
                            "exchange": "binance",
                            "symbol": "ETHUSDT",
                            "type": "linear",
                            "side": "buy",
                            "orderId": "def456",
                            "price": 2953.26,
                            "priceUpdated": True,
                            "originalOrderId": None
                        }
                    }
                },
                {
                    "description": "TWAP 執行記錄範例",
                    "value": {
                        "ts": 1766508307833,
                        "mode": "twap",
                        "strategyId": "twap_6168e245",
                        "pairId": None,
                        "twapId": "twap_6168e245",
                        "totalTriggers": 1,
                        "status": "success",
                        "reason": None,
                        "error": None,
                        "qty": 0.01,
                        "spread": -0.4,
                        "spreadPercent": -0.013548,
                        "totalAmount": 0.01,
                        "orderCount": 1,
                        "threshold": None,
                        "intervalMs": 10000,
                        "isRollback": False,
                        "leg1": {
                            "exchange": "bybit",
                            "symbol": "ETHUSDT",
                            "type": "spot",
                            "side": "buy",
                            "orderId": "xyz789",
                            "price": 2952.67,
                            "priceUpdated": True,
                            "originalOrderId": None
                        },
                        "leg2": {
                            "exchange": "binance",
                            "symbol": "ETHUSDT",
                            "type": "spot",
                            "side": "sell",
                            "orderId": "uvw012",
                            "price": 2952.27,
                            "priceUpdated": True,
                            "originalOrderId": None
                        }
                    }
                }
            ]
        }

    @classmethod
    def from_pair_dict(cls, record: dict) -> "UnifiedExecutionRecord":
        """從 Pairs 舊格式字典轉換為統一格式
        
        Args:
            record: 舊格式的 Pairs 執行記錄
            
        Returns:
            統一格式的執行記錄
        """
        pair_id = record.get("pairId")
        
        return cls(
            ts=record.get("ts", int(time.time() * 1000)),
            mode="pair",
            strategyId=pair_id,
            pairId=pair_id,
            twapId=None,
            
            totalTriggers=record.get("totalTriggers", 1),
            
            status=record.get("status", "success"),
            reason=record.get("reason"),
            error=record.get("error"),
            
            qty=record.get("qty", 0),
            spread=record.get("spread"),
            spreadPercent=record.get("spreadPercent"),
            
            totalAmount=record.get("maxExecs", 1) * record.get("qty", 0),
            orderCount=record.get("maxExecs", 1),
            threshold=record.get("threshold"),
            intervalMs=None,
            
            isRollback=record.get("isRollback", False),
            
            leg1=ExecutionLeg(**record["leg1"]) if record.get("leg1") else None,
            leg2=ExecutionLeg(**record["leg2"]) if record.get("leg2") else None
        )

    @classmethod
    def from_twap_dict(cls, record: dict) -> "UnifiedExecutionRecord":
        """從 TWAP 舊格式字典轉換為統一格式
        
        Args:
            record: 舊格式的 TWAP 執行記錄
            
        Returns:
            統一格式的執行記錄
        """
        plan_id = record.get("planId")
        slice_index = record.get("sliceIndex", 0)
        
        return cls(
            ts=record.get("ts", int(time.time() * 1000)),
            mode="twap",
            strategyId=plan_id,
            pairId=None,
            twapId=plan_id,
            
            # TWAP 的 totalTriggers = sliceIndex + 1（從1開始）
            totalTriggers=slice_index + 1,
            
            status=record.get("status", "success"),
            reason=record.get("reason"),
            error=record.get("error"),
            
            qty=record.get("qty", 0),
            spread=record.get("spread"),
            spreadPercent=record.get("spreadPercent"),
            
            totalAmount=record.get("totalAmount", 0),
            orderCount=record.get("orderCount", 1),
            threshold=None,
            intervalMs=record.get("intervalMs"),
            
            isRollback=record.get("isRollback", False),
            
            leg1=ExecutionLeg(**record["leg1"]) if record.get("leg1") else None,
            leg2=ExecutionLeg(**record["leg2"]) if record.get("leg2") else None
        )

    def to_dict(self) -> dict:
        """轉換為字典格式（用於寫入 JSONL）
        
        Returns:
            字典格式的執行記錄
        """
        result = {
            "ts": self.ts,
            "mode": self.mode,
            "strategyId": self.strategyId,
            "pairId": self.pairId,
            "twapId": self.twapId,
            
            "totalTriggers": self.totalTriggers,
            
            "status": self.status,
            "reason": self.reason,
            "error": self.error,
            
            "qty": self.qty,
            "spread": self.spread,
            "spreadPercent": self.spreadPercent,
            
            "totalAmount": self.totalAmount,
            "orderCount": self.orderCount,
            "threshold": self.threshold,
            "intervalMs": self.intervalMs,
            
            "isRollback": self.isRollback,
            
            "leg1": self.leg1.model_dump() if self.leg1 else None,
            "leg2": self.leg2.model_dump() if self.leg2 else None
        }
        
        return result

File: client/src/types/execution.ts
/**
 * 統一執行記錄型別定義 V2
 * 適用於 TWAP 和 Pairs 策略
 * 與後端 UnifiedExecutionRecord 模型保持一致
 */

// 基礎型別
export type ExecutionMode = "pair" | "twap";
export type ExecutionStatus = "success" | "failed" | "cancelled" | "rolled_back";
export type ExchangeName = "bybit" | "binance" | "okx" | "bitget";
export type TradeType = "spot" | "linear" | "inverse";
export type TradeSide = "buy" | "sell";

/**
 * 執行腿的統一結構
 */
export interface ExecutionLeg {
  exchange: ExchangeName;
  symbol: string;
  type: TradeType;
  side: TradeSide;
  orderId: string | null;
  price: number | null;
  priceUpdated: boolean;
  originalOrderId?: string | null;  // 回滾時使用
}

/**
 * 統一的執行記錄格式（TWAP 和 Pairs 共用）
 */
export interface UnifiedExecutionRecord {
  // 基本資訊
  ts: number;                      // 時間戳（毫秒）
  mode: ExecutionMode;             // 策略類型
  strategyId: string;              // 策略ID（統一識別用）
  pairId: string | null;           // Pairs策略ID（mode=pair時使用）
  twapId: string | null;           // TWAP策略ID（mode=twap時使用）
  
  // 執行資訊
  totalTriggers: number;           // 第幾次執行（從1開始，pair和twap共用）
  
  // 狀態資訊
  status: ExecutionStatus;         // 執行狀態
  reason: string | null;           // 失敗或取消原因
  error: string | null;            // 錯誤訊息
  
  // 數量和價格資訊
  qty: number;                     // 本次執行數量
  spread: number | null;           // 價差
  spreadPercent: number | null;    // 價差百分比
  
  // 策略配置資訊
  totalAmount: number;             // 預期總數量
  orderCount: number;              // 總執行次數/最大執行次數
  threshold: number | null;        // 觸發閾值（pair用，twap為null）
  intervalMs: number | null;       // 執行間隔毫秒（twap用，pair為null）
  
  // 回滾資訊
  isRollback: boolean;             // 是否為回滾記錄
  
  // 執行腿資訊
  leg1: ExecutionLeg | null;       // 第一腿執行結果
  leg2: ExecutionLeg | null;       // 第二腿執行結果
}

/**
 * 型別守衛：檢查是否為統一執行記錄
 */
export function isUnifiedExecutionRecord(obj: any): obj is UnifiedExecutionRecord {
  return obj &&
         typeof obj === 'object' &&
         typeof obj.ts === 'number' &&
         typeof obj.mode === 'string' &&
         (obj.mode === 'pair' || obj.mode === 'twap') &&
         typeof obj.strategyId === 'string' &&
         typeof obj.totalTriggers === 'number' &&
         typeof obj.status === 'string' &&
         typeof obj.qty === 'number' &&
         typeof obj.totalAmount === 'number' &&
         typeof obj.orderCount === 'number' &&
         typeof obj.isRollback === 'boolean';
}

/**
 * 型別守衛：檢查是否為 Pairs 執行記錄
 */
export function isPairExecutionRecord(obj: UnifiedExecutionRecord): boolean {
  return obj.mode === 'pair' && obj.pairId !== null;
}

/**
 * 型別守衛：檢查是否為 TWAP 執行記錄
 */
export function isTwapExecutionRecord(obj: UnifiedExecutionRecord): boolean {
  return obj.mode === 'twap' && obj.twapId !== null;
}

/**
 * 工具函數：計算執行進度百分比
 */
export function getExecutionProgress(record: UnifiedExecutionRecord): number {
  return Math.min((record.totalTriggers / record.orderCount) * 100, 100);
}

/**
 * 工具函數：判斷是否已完成
 */
export function isExecutionCompleted(record: UnifiedExecutionRecord): boolean {
  return record.totalTriggers >= record.orderCount;
}

/**
 * 工具函數：取得策略顯示名稱
 */
export function getStrategyDisplayName(record: UnifiedExecutionRecord): string {
  if (record.mode === 'pair') {
    return `Pairs: ${record.pairId}`;
  } else {
    return `TWAP: ${record.twapId}`;
  }
}

/**
 * 工具函數：格式化執行狀態
 */
export function formatExecutionStatus(status: ExecutionStatus): string {
  const statusMap: Record<ExecutionStatus, string> = {
    'success': '成功',
    'failed': '失敗',
    'cancelled': '已取消',
    'rolled_back': '已回滾'
  };
  return statusMap[status] || status;
}

/**
 * 工具函數：取得狀態顏色
 */
export function getStatusColor(status: ExecutionStatus): string {
  const colorMap: Record<ExecutionStatus, string> = {
    'success': 'green',
    'failed': 'red',
    'cancelled': 'orange',
    'rolled_back': 'purple'
  };
  return colorMap[status] || 'gray';
}

/**
 * 聚合函數：按 strategyId 分組執行記錄
 */
export function groupRecordsByStrategy(
  records: UnifiedExecutionRecord[]
): Record<string, UnifiedExecutionRecord[]> {
  return records.reduce((acc, record) => {
    const key = record.strategyId;
    if (!acc[key]) {
      acc[key] = [];
    }
    acc[key].push(record);
    return acc;
  }, {} as Record<string, UnifiedExecutionRecord[]>);
}

/**
 * 聚合函數：計算策略的總成交數量
 */
export function calculateTotalExecutedQty(
  records: UnifiedExecutionRecord[],
  strategyId: string
): { leg1TotalQty: number; leg2TotalQty: number } {
  const strategyRecords = records.filter(
    r => r.strategyId === strategyId && r.status === 'success'
  );
  
  return {
    leg1TotalQty: strategyRecords.reduce((sum, r) => sum + r.qty, 0),
    leg2TotalQty: strategyRecords.reduce((sum, r) => sum + r.qty, 0)
  };
}

/**
 * 過濾函數：取得成功的執行記錄
 */
export function getSuccessfulRecords(
  records: UnifiedExecutionRecord[]
): UnifiedExecutionRecord[] {
  return records.filter(r => r.status === 'success' && !r.isRollback);
}

/**
 * 過濾函數：取得失敗的執行記錄
 */
export function getFailedRecords(
  records: UnifiedExecutionRecord[]
): UnifiedExecutionRecord[] {
  return records.filter(r => r.status === 'failed');
}

/**
 * 過濾函數：取得回滾的執行記錄
 */
export function getRollbackRecords(
  records: UnifiedExecutionRecord[]
): UnifiedExecutionRecord[] {
  return records.filter(r => r.isRollback);
}

/**
 * 向後兼容：將舊格式轉換為統一格式（Pairs）
 */
export function convertLegacyPairRecord(legacy: any): UnifiedExecutionRecord {
  return {
    ts: legacy.ts || 0,
    mode: 'pair',
    strategyId: legacy.pairId || legacy.strategyId,
    pairId: legacy.pairId,
    twapId: null,
    totalTriggers: legacy.totalTriggers || legacy.executionIndex || 1,
    status: legacy.status || 'success',
    reason: legacy.reason || null,
    error: legacy.error || null,
    qty: legacy.qty || legacy.sliceQty || 0,
    spread: legacy.spread || null,
    spreadPercent: legacy.spreadPercent || null,
    totalAmount: legacy.totalAmount || (legacy.maxExecs || 1) * (legacy.qty || 0),
    orderCount: legacy.orderCount || legacy.maxExecs || 1,
    threshold: legacy.threshold || null,
    intervalMs: null,
    isRollback: legacy.isRollback || false,
    leg1: legacy.leg1 || null,
    leg2: legacy.leg2 || null
  };
}

/**
 * 向後兼容：將舊格式轉換為統一格式（TWAP）
 */
export function convertLegacyTwapRecord(legacy: any): UnifiedExecutionRecord {
  const sliceIndex = legacy.sliceIndex || 0;
  
  return {
    ts: legacy.ts || 0,
    mode: 'twap',
    strategyId: legacy.planId || legacy.twapId || legacy.strategyId,
    pairId: null,
    twapId: legacy.planId || legacy.twapId,
    totalTriggers: legacy.executionIndex || (sliceIndex + 1),
    status: legacy.status || 'success',
    reason: legacy.reason || null,
    error: legacy.error || null,
    qty: legacy.qty || legacy.sliceQty || 0,
    spread: legacy.spread || null,
    spreadPercent: legacy.spreadPercent || null,
    totalAmount: legacy.totalAmount || 0,
    orderCount: legacy.orderCount || 0,
    threshold: null,
    intervalMs: legacy.intervalMs || null,
    isRollback: legacy.isRollback || false,
    leg1: legacy.leg1 || null,
    leg2: legacy.leg2 || null
  };
}

/**
 * 常數定義
 */
export const SUPPORTED_EXECUTION_MODES: ExecutionMode[] = ['pair', 'twap'];
export const SUPPORTED_EXECUTION_STATUSES: ExecutionStatus[] = [
  'success',
  'failed',
  'cancelled',
  'rolled_back'
];

